File: src/main/java/skywolf46/adl/abstraction/AbstractDataLoader.java
package skywolf46.adl.abstraction;

import skywolf46.adl.AsyncDataLoader;
import skywolf46.adl.interfaces.IDataQueueable;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

public abstract class AbstractDataLoader implements IDataQueueable {
    private AtomicInteger saveCountdown = new AtomicInteger(60);
    private final AtomicBoolean isLoaded = new AtomicBoolean(false);
    private AbstractDataTask task = null;
    private List<Runnable> rs = new ArrayList<>();

    public void load() {
        unload();
        loadData();
        synchronized (isLoaded) {
            isLoaded.set(true);
            for (Runnable ra : rs)
                ra.run();
            rs.clear();
        }
    }

    public void unload() {
        isLoaded.set(false);
    }

    protected synchronized void setLoaded(boolean isLoaded) {
        this.isLoaded.set(isLoaded);
        if(isLoaded)
            runQueued();
    }

    public final void queue(Runnable r) {
        synchronized (isLoaded) {
            if (!isLoaded()) {
                rs.add(r);
                return;
            }
        }
        r.run();
    }

    public void runQueued() {
        synchronized (isLoaded) {
            for (Runnable r : rs)
                r.run();
            rs.clear();
        }
    }

    public boolean isLoaded() {
        return isLoaded.get();
    }

    protected abstract void loadData();


    public abstract AbstractDataSnapshot<?> createSnapshot();

    public final void mark(Class<? extends AbstractDataTask> cls, int countdown) {
//        System.out.println("Marked!");
        if (task == null || !task.getClass().equals(cls)) {
            if (AsyncDataLoader.getTaskFactory(cls) == null) {
                return;
            }
            task = AsyncDataLoader.getTaskFactory(cls).createSaveTask(this);
        }
        saveCountdown.set(countdown);
    }

    public final void load(Class<? extends AbstractDataTask> cls) {
        AsyncDataLoader.getTaskFactory(cls).createLoadTask(this).load();
    }

    public AtomicInteger getSaveCountdown() {
        return saveCountdown;
    }
}

File: src/main/kotlin/skywolf46/adl/abstraction/AbstractDataLoader.kt
package skywolf46.adl.abstraction

import org.bukkit.Bukkit
import skywolf46.adl.AsyncDataLoader
import skywolf46.adl.abstraction.tasks.SchedulerBasedDataSaveTask
import skywolf46.extrautility.util.ifFalse
import skywolf46.extrautility.util.ifTrue
import skywolf46.extrautility.util.runNonNull
import java.util.concurrent.atomic.AtomicBoolean

abstract class AbstractDataLoader<T, F : AbstractDataSnapshot<T>> : IDataQueueable{
    var isLoaded: AtomicBoolean = AtomicBoolean(false)
        private set
    val cache: MutableList<AbstractDataLoader<T, F>.() -> Unit> = ArrayList()
    private var task: AbstractDataTask? = null

    abstract fun loadData(data: T)

    fun asyncLoadAndExecute(data: T, lambda: () -> Void) {
        Bukkit.getScheduler().runTaskAsynchronously(AsyncDataLoader.inst) {
            loadData(data)
            Bukkit.getScheduler().scheduleSyncDelayedTask(AsyncDataLoader.inst) {
                lambda()
            }
        }
    }

    abstract fun snapshot(): F

    fun forceSave(afterRun: () -> Unit) {
        runNonNull(task) {
            stopTimer()
            task = null
        }
        snapshot().saveSnapshotAndRun { afterRun() }
    }

    fun triggerSave(cls: Class<out AbstractDataTask>, time: Int) {
        task = task ?: AsyncDataLoader.inst?.of(cls)?.createSaver(this)?.apply {
            startTimer()
        }
        task?.resetTrigger(time)
    }

    fun triggerSave(time: Int) = triggerSave(SchedulerBasedDataSaveTask::class.java, time)

    fun triggerSave() = triggerSave(60)

    fun destroyTask() {
        task = null
    }

    protected fun runAllCaches() {
        val lst: MutableList<AbstractDataLoader<T, F>.() -> Unit> = ArrayList()
        synchronized(this) {
            lst.addAll(cache)
            cache.clear()
        }
        for (x in lst)
            x(this)
    }

}

@Suppress("UNCHECKED_CAST")
fun <T, F : AbstractDataSnapshot<T>, S : AbstractDataLoader<T, F>> S.runIfLoaded(unit: S.() -> Unit): S {
    synchronized(this) {
        isLoaded.get().ifFalse {
            cache.add(unit as AbstractDataLoader<T, F>.() -> Unit)
        }.ifTrue {
            unit()
        }
    }
    return this
}

